### DISCUSSION TITLE
Question about injecting and generating new classes

### DISCUSSION THREAD

**Discussion by anonymousgroundhog** [soot-oss/SootUp#1057] [General]
I want to know if it is possible to inject a new class.

For example, if I were to use the following code:
```java
AnalysisInputLocation inputLocation = new ApkAnalysisInputLocation(pathToAPK, SourceType.Application);

        JavaView view = new JavaView(inputLocation);

        // get all classes
        Collection<JavaSootClass> classes = view.getClasses();
```
I want to insert an empty class.

Also, can sootup output the jimple to a folder similar to the old version of soot?
Any help would be much appreciated.

Thanks.

**Comment by stschott** [comment]
Hello, in regards to creating a and inserting a new SootClass into the View please have a look at the `MutableJavaView`. You can find an example of its usage [here](https://github.com/soot-oss/SootUp-Examples/blob/main/MutatingSootClassExample/src/main/java/sootup/examples/MutatingSootClass.java).

To print the Jimple of a SootClass into a file you can use the `JimplePrinter`. See the following code snippet:
```java
JimplePrinter jimplePrinter = new JimplePrinter();
FileWriter fileWriter = new FileWriter("example.jimple");
PrintWriter printWriter = new PrintWriter(fileWriter);
jimplePrinter.printTo(sootClassToPrint, printWriter);
printWriter.close();
```

**Reply by anonymousgroundhog** [reply]
Thank you this makes more sense now.

Also, how could you get soot to output an APK file using the example code demonstrated?

**Reply by anonymousgroundhog** [reply]
Also, could you show a demo of how to generate the sootClassToPrint with an empty class?

For example, in old soot, you could do `SootClass sootClassToPrint = new SootClass("HelloWorld", Modifier.PUBLIC);`

**Reply by stschott** [reply]
> Thank you this makes more sense now.
> 
> Also, how could you get soot to output an APK file using the example code demonstrated?

Unfortunately this is not possible at the moment.


> Also, could you show a demo of how to generate the sootClassToPrint with an empty class?
> 
> For example, in old soot, you could do `SootClass sootClassToPrint = new SootClass("HelloWorld", Modifier.PUBLIC);`

You can create a new emtpy JavaSootClass like this:
```java
    JavaSootClass mainClass =
        new JavaSootClass(
            new OverridingJavaClassSource(
                new EagerInputLocation(),
                null,
                "HelloWorld", new PackageName(""),
                null,
                Collections.emptySet(),
                null,
                Collections.emptySet(),
                Collections.emptySet(),
                NoPositionInformation.getInstance(),
                EnumSet.of(ClassModifier.PUBLIC),
                Collections.emptyList(),
                Collections.emptyList(),
                Collections.emptyList()),
            SourceType.Application);
```

Have a look at the OverridingClassSource [Javadoc](https://soot-oss.github.io/SootUp/apidocs/sootup/java/core/OverridingJavaClassSource.html) to see how to modify specific parts of the class.

**Reply by anonymousgroundhog** [reply]
Ok, thank you so much for your assistance @stschott.
 
Having the ability to generate a dex output would be much needed, especially for those of us doing malware behavioral analysis for research purposes.
I don't sound like I am complaining, but I just wanted to note this down.

**Reply by stschott** [reply]
> Having the ability to generate a dex output would be much needed, especially for those of us doing malware behavioral analysis for research purposes. I don't sound like I am complaining, but I just wanted to note this down.

We started working on it. However, it will still take some time until it will be ready.

**Reply by anonymousgroundhog** [reply]
Thank you. Is this under way?
I know from experience that updating others' code can be quite a task.

Much appreciated.
Is there an expected date or time for this feature to be released?

**Reply by stschott** [reply]
It will still take a while until it is ready. Sometime around spring - summer 2025.
